### DISCUSSION TITLE
"Zero or One" rule

### DISCUSSION THREAD

**Discussion by tomdavsmi** [foonathan/lexy#47] [Q&A] [answered]
Hi! 

I've been trying to get to grips with lexy by writing a chemical formula parser, and I can't quite 
figure out how to emulate the ? construct from regular expressions (i.e. zero or one). I would need this to make "H" or "Fe" to parse as correct, but "Hee" not. 

I'm quite new to this kind of thing so there is absolutely a chance that I've just glossed over/not understood the right bit of the documentation. 

Thank you!

**Comment by foonathan** [comment, marked as answer]
You can use either [`dsl::if_()`](https://lexy.foonathan.net/reference/dsl/if/#if_) or [`dsl::opt()`](https://lexy.foonathan.net/reference/dsl/option/#opt) for that. The latter gives you a value when the rule didn't match, the former doesn't.

Although, if you're having a fixed list of symbols, it might make sense to use [`dsl::symbol`](https://lexy.foonathan.net/reference/dsl/symbol/): you specify a mapping of e.g. element names to enum values in a table, then read one of those.

**Reply by tomdavsmi** [reply]
Ahh okay, I'll go and have a play around to try and get more familiar with how branching works. And you're absolutely right, a fixed list of symbols is a much better idea!

Thanks again.
